Test BaseModel and fix bug

jugglinmike opened the pull request:
In order to fix gh-5, I started by backing the current behavior of the `BaseModel` class with unit tests. That work is concentrated in the first commit of this patch and hopefully guards against regressions for functional changes to the class. The second commit modifies framework code in order to resolve gh-5.

carldanley commented on test/tests/models/BaseModel.php:
```diff
@@ -0,0 +1,674 @@
+require_once(__DIR__ . '/../../../utils/models/BaseModel.php');
+require_once(__DIR__ . '/../../AEFrameworkTestCase.php');
+
+function defaultFunction()
```
Please remove this `defaultFunction` from the global namespace and include it inline via a closure instead.

carldanley commented on utils/database.php:
```diff
@@ -43,6 +43,21 @@ class Database implements DatabaseInterface {
+   * Reference:
+   * https://secure.php.net/manual/en/mysqli-stmt.bind-param.php
+   */
+  const MYSQLI_BIND_PARAM_TYPES = array(
```
:+1: for this documentation.

jugglinmike commented:
@carldanley I've attempted to follow your suggestion, but it isn't pretty. I also didn't really care for the global function, but this new approach is very awkward. My initial thought was to declare the closure within the `_db_schema` class property itself, but I can't define a closure as a class property. I assume there is some language feature that I'm missing. Any thoughts?

bmac commented on test/tests/models/BaseModel.php:
```diff
@@ -0,0 +1,676 @@
+    $this->assertSame($this->m->sanitize_db_field_value('test_attr3', false), '');
+    $this->assertSame($this->m->sanitize_db_field_value('test_attr3', null), '');
+  }
+}
```
Yay. Tests \o/

carldanley commented:
@jugglinmike - another alternative is to create the method on the class itself and then, instead of referencing it by `string`, reference it by `array` (ie - `array($class, $method)`. Then, you can call this function by performing `call_user_func_array()`. Maybe that feels less icky?

carldanley commented:
Also, if you go that method, be sure to make use of [`is_callable()`](http://php.net/manual/en/function.is-callable.php) to ensure that `call_user_func_array()` won't fail.

carldanley commented:
Other than that closure stuff, this looks awesome. code :+1: tests :+1:

jugglinmike commented:
Thanks, @carldanley! I'm not sure how to proceed in this case. It looks like [the only real use of this feature is to specify global functions via string values](https://github.com/wfmu/ae-api/search?utf8=%E2%9C%93&q=defaultFunction). We're having trouble testing the other methods of defining `defaultFunction`, but application code would have the same problems.

This makes me think that the original approach (specifying a function via a string) is best not because it is easy, but because it is the behavior that is actually relied upon. I just spent too much time trying to figure out a way to undefine a global function (which we could do in [PHPUnit's `tearDownAfterClass` method](https://phpunit.de/manual/current/en/fixtures.html#fixtures.more-setup-than-teardown)), only to learn that:

> PHP does not support function overloading, nor is it possible to undefine or redefine previously-declared functions.

(source: http://www.php.net/manual/en/functions.user-defined.php)

But then I had a better idea! A static method on the `Model` class we're defining for testing purposes. I've pushed up a commit with this new approach--what do you think?

carldanley commented:
I should have looked more in depth earlier, apologies for that. What I'm describing should already be handled via https://github.com/wfmu/ae-framework/blob/master/utils/models/BaseModel.php#L153-L163.

The only thing you'd need to do is define a method (within the class), like this:

```php
class Thing {
  public function defaultFunction() {
    return 'hey';
  }
}
```

Then, for the value of `defaultFunction`, you can specify 1 of the following:
* `array('Thing', 'defaultFunction');`
* `array($this, 'defaultFunction');`
* `array(self, 'defaultFunction');`

I think any of those values should work. Does this make sense?

jugglinmike commented:
@carldanley That makes sense, but since it's a separate behavior, I think we should have a separate test for it. I'll get to work updating the tests to exercise that code path via a new `test_attr5`, but in the mean time, do you think the current approach works for the "non-array `defaultFunction`" case?

carldanley commented:
Yeah, I think that should work great. Thanks!

jugglinmike commented:
I think the `call_user_func_array` code branch is incorrect. [The PHP docs
describe it as a function with arity 2](https://secure.php.net/manual/en/function.call-user-func-array.php), but [we're calling it with a single argument](https://github.com/wfmu/ae-framework/blob/8c235f78c7b88fe9d047f30ee7326d76d95fb475/utils/models/BaseModel.php#L156).
What we need is a two-dimensional array of length two: the first element
specifies the function (and may itself be an array), and the second specifies
the arguments to that function (and must be an array).

That said, I think we might want to remove it instead of fixing it. Here's why:

The `defaultFunction` must be specified declaratively in a class definition, so
it's unlikely that consumers will need to specify variable arguments. The only
time this would be useful is if consumers had some generic, global
"defaultFunction factory" that could generate valid `defaultFunction`s based
on some set of variable parameters. This global function could be re-used
across class definitions like so:

```php
class M1 extends BaseModel {
  protected $_db_schema = array(
    'test_attr5' => array(
      'type' => 'int',
      'defaultFunction' => array(
        'defaultFunctionFactory',
        array('default for M1', 'some other parameter')
      )
    )
  );
}
```

and elsewhere:

```php
class M2 extends BaseModel {
  protected $_db_schema = array(
    'test_attr5' => array(
      'type' => 'int',
      'defaultFunction' => array(
        'defaultFunctionFactory',
        array('default for M2', 'some other parameter')
      )
    )
  );
}
```

But that seems like a high amount of abstraction for little benefit. As I
mentioned earlier, no application code is currently relying on that code path.
This suggests its limited usefulness, and it also simplifies the prospect of
removing the feature (no change to `ae-api` would be required).

What would you like to do?

jugglinmike commented:
@carldanley I've added another commit as per our discussion in chat

carldanley commented:
Talked with everyone about this on the call but this is good to go, code :+1: tests :+1: - @jugglinmike created an issue somewhere for that (i think?). do you have that link handy? didn't see it on issues for ae-framework.

jugglinmike commented:
Rebased, squashed, and merged to `master`. I only just now created that follow-up issue; you can find it here: https://github.com/wfmu/ae-framework/issues/27

Thanks, @carldanley!
